File: chapter-07-automation/python/01-webhooks/webhook_manager.py
#!/usr/bin/env python3
# webhook_manager.py
# Ejemplo: Gestionar suscripciones webhook con Microsoft Graph

import uuid
from datetime import datetime, timedelta
from msgraph import GraphServiceClient


class WebhookManager:
    def __init__(self, graph_client: GraphServiceClient, notification_url: str):
        self.graph_client = graph_client
        self.notification_url = notification_url

    async def create_drive_subscription(self, drive_id: str):
        print(f"🔔 Creando suscripción para Drive: {drive_id}")

        subscription = {
            "resource": f"/drives/{drive_id}/root",
            "changeType": "created,updated,deleted",
            "notificationUrl": self.notification_url,
            "clientState": str(uuid.uuid4()),
            "expirationDateTime": (datetime.utcnow() + timedelta(days=2)).isoformat()
        }

        result = await self.graph_client.subscriptions.post(subscription)
        print(f"   ✅ Suscripción: {result.id}")
        return result

    async def renew_subscription(self, subscription_id: str):
        patch = {
            "expirationDateTime": (datetime.utcnow() + timedelta(days=2)).isoformat()
        }

        result = await self.graph_client.subscriptions.by_subscription_id(
            subscription_id
        ).patch(patch)
        print(f"🔄 Renovada. Expira: {result.expiration_date_time}")
        return result

    async def delete_subscription(self, subscription_id: str):
        await self.graph_client.subscriptions.by_subscription_id(
            subscription_id
        ).delete()
        print(f"🗑️ Suscripción {subscription_id} eliminada")


# Uso
# manager = WebhookManager(graph_client, "https://tu-endpoint.com/webhook")
# await manager.create_drive_subscription("drive-id")

File: chapter-07-automation/csharp/01-webhooks/WebhookManager.cs
// WebhookManager.cs
// Example: Create and manage webhook subscriptions in Microsoft Graph

using Microsoft.Graph;
using Microsoft.Graph.Models;

namespace AutomationExamples;

public class WebhookManager
{
    private readonly GraphServiceClient _graphClient;
    private readonly string _notificationUrl;

    public WebhookManager(GraphServiceClient graphClient, string notificationUrl)
    {
        _graphClient = graphClient;
        _notificationUrl = notificationUrl;
    }

    public async Task<Subscription> CreateDriveSubscriptionAsync(string driveId)
    {
        Console.WriteLine($"🔔 Creating subscription for Drive: {driveId}");

        var subscription = new Subscription
        {
            Resource = $"/drives/{driveId}/root",
            ChangeType = "created,updated,deleted",
            NotificationUrl = _notificationUrl,
            ClientState = Guid.NewGuid().ToString(),
            ExpirationDateTime = DateTime.UtcNow.AddDays(2)
        };

        var created = await _graphClient.Subscriptions.PostAsync(subscription);
        if (created == null) throw new InvalidOperationException("Graph did not return the created subscription.");

        Console.WriteLine($"   ✅ Subscription: {created.Id}");
        Console.WriteLine($"   Expires: {created.ExpirationDateTime}");

        return created;
    }

    public async Task RenewSubscriptionAsync(string subscriptionId)
    {
        var patch = new Subscription
        {
            ExpirationDateTime = DateTime.UtcNow.AddDays(2)
        };

        var renewed = await _graphClient.Subscriptions[subscriptionId].PatchAsync(patch);
        if (renewed != null)
        {
            Console.WriteLine($"🔄 Renewed. New expiration: {renewed.ExpirationDateTime}");
        }
    }

    public async Task DeleteSubscriptionAsync(string subscriptionId)
    {
        await _graphClient.Subscriptions[subscriptionId].DeleteAsync();
        Console.WriteLine($"🗑️ Subscription {subscriptionId} deleted");
    }
}
